fieldservice_sale/models/sale_order.py

```diff
@@ -0,0 +1,213 @@
+# Copyright (C) 2021 Brian McMaster
+# Copyright (C) 2021 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    fsm_location_id = fields.Many2one(
+        "fsm.location",
+        string="Service Location",
+        help="SO Lines generating a FSM order will be for this location",
+    )
+    fsm_order_ids = fields.Many2many(
+        "fsm.order",
+        compute="_compute_fsm_order_ids",
+        string="Field Service orders associated to this sale",
+    )
+    fsm_order_count = fields.Integer(
+        string="FSM Orders", compute="_compute_fsm_order_ids"
+    )
+
+    @api.depends("order_line")
+    def _compute_fsm_order_ids(self):
+        for order in self:
+            orders = self.env["fsm.order"]
+            orders |= self.env["fsm.order"].search(
+                [("sale_line_id", "in", order.order_line.ids)]
+            )
+            orders |= self.env["fsm.order"].search([("sale_id", "=", order.id)])
+            order.fsm_order_ids = orders
+            order.fsm_order_count = len(order.fsm_order_ids)
+
+    @api.onchange("partner_id")
+    def onchange_partner_id(self):
+        """
+        Autofill the Sale Order's FS location with the partner_id,
+        the partner_shipping_id or the partner_id.commercial_partner_id if
+        they are FS locations.
+        """
+        super(SaleOrder, self).onchange_partner_id()
+        domain = [
+            "|",
+            "|",
+            ("partner_id", "=", self.partner_id.id),
+            ("partner_id", "=", self.partner_shipping_id.id),
+            ("partner_id", "=", self.partner_id.commercial_partner_id.id),
+        ]
+        if self.partner_id.fsm_location:
+            domain = [("partner_id", "=", self.partner_id.id)]
+        location_ids = self.env["fsm.location"].search(domain)
+        self.fsm_location_id = location_ids and location_ids[0] or False
+
+    def _field_create_fsm_order_prepare_values(self):
+        self.ensure_one()
+        lines = self.order_line.filtered(
+            lambda sol: sol.product_id.field_service_tracking == "sale"
+        )
+        templates = lines.mapped("product_id.fsm_order_template_id")
+        note = ""
+        hours = 0.0
+        categories = self.env["fsm.category"]
+        for template in templates:
+            note += template.instructions or ""
+            hours += template.hours
```

Comment:
@Freni-OSI This should be template.duration

Reply:
@osi-scampbell Changed in Latest PR #761.